File: examples/extract-download-images/rest/python/images.py
# Extracts all <img> src URLs from a page via the Browserless /scrape endpoint,
# then downloads each image to an images/ directory.
#
# Install: pip install requests
# Run:     python images.py

import os
import requests
from urllib.parse import urlparse

TOKEN = 'YOUR_API_TOKEN_HERE'

scrape_res = requests.post(
    f'https://production-sfo.browserless.io/scrape?token={TOKEN}',
    json={
        'url': 'https://example.com',
        'elements': [{'selector': 'img', 'timeout': 5000}],
    },
)

data = scrape_res.json()['data']

# src is nested inside each result's attributes array.
image_urls = [
    attr['value']
    for result in data[0]['results']
    for attr in result['attributes']
    if attr['name'] == 'src' and attr['value'].startswith('http')
]

print(f'Found {len(image_urls)} images')
os.makedirs('images', exist_ok=True)

for i, url in enumerate(image_urls):
    res = requests.get(url)
    ext = os.path.splitext(urlparse(url).path)[1] or '.jpg'
    filename = f'images/image-{i}{ext}'
    with open(filename, 'wb') as f:
        f.write(res.content)
    print(f'Saved {filename}')

File: examples/extract-download-images/rest/nodejs/images.mjs
// Extracts all <img> src URLs from a page via the Browserless /scrape endpoint,
// then downloads each image to an images/ directory.
// Requires Node.js 18+ for the native fetch API. No extra packages needed.
//
// Run: node images.mjs

import fs from 'fs';
import path from 'path';

const TOKEN = 'YOUR_API_TOKEN_HERE';

const scrapeRes = await fetch(
  `https://production-sfo.browserless.io/scrape?token=${TOKEN}`,
  {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      url: 'https://example.com',
      elements: [{ selector: 'img', timeout: 5000 }],
    }),
  }
);

const { data } = await scrapeRes.json();

// src is nested inside each result's attributes array.
const imageUrls = data[0].results
  .flatMap((r) => r.attributes)
  .filter((a) => a.name === 'src' && a.value.startsWith('http'))
  .map((a) => a.value);

console.log(`Found ${imageUrls.length} images`);
fs.mkdirSync('images', { recursive: true });

for (const [i, url] of imageUrls.entries()) {
  const res = await fetch(url);
  const buf = Buffer.from(await res.arrayBuffer());
  const ext = path.extname(new URL(url).pathname) || '.jpg';
  fs.writeFileSync(`images/image-${i}${ext}`, buf);
  console.log(`Saved image-${i}${ext}`);
}
